src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -4552,31 +4589,101 @@ private BoundExpression BindCollectionLiteralExpression(CollectionCreationExpres
             {
                 builder.Add(bindElement(element, diagnostics));
             }
-            return new BoundUnconvertedCollectionLiteralExpression(syntax, builder.ToImmutableAndFree(), this);
+
+            var initializers = builder.ToImmutableAndFree();
+            TypeSymbol? collectionType = InferCollectionLiteralType(initializers);
+            return new BoundUnconvertedCollectionLiteralExpression(syntax, initializers, this, type: collectionType);
 
             BoundExpression bindElement(CollectionElementSyntax syntax, BindingDiagnosticBag diagnostics)
             {
-                switch (syntax)
+                return syntax switch
                 {
-                    case ExpressionElementSyntax expressionElementSyntax:
-                        return BindValue(expressionElementSyntax.Expression, diagnostics, BindValueKind.RValue);
+                    ExpressionElementSyntax expressionElementSyntax => BindValue(expressionElementSyntax.Expression, diagnostics, BindValueKind.RValue),
+                    DictionaryElementSyntax dictionaryElementSyntax => bindDictionaryElement(dictionaryElementSyntax, diagnostics),
+                    SpreadElementSyntax spreadElementSyntax => bindSpreadElement(spreadElementSyntax, diagnostics),
+                    _ => throw ExceptionUtilities.UnexpectedValue(syntax.Kind())
+                };
+            }
 
-                    case DictionaryElementSyntax dictionaryElementSyntax:
-                        _ = BindValue(dictionaryElementSyntax.KeyExpression, diagnostics, BindValueKind.RValue);
-                        _ = BindValue(dictionaryElementSyntax.ValueExpression, diagnostics, BindValueKind.RValue);
-                        Error(diagnostics, ErrorCode.ERR_CollectionLiteralElementNotImplemented, syntax);
-                        return BadExpression(syntax);
+            BoundExpression bindDictionaryElement(DictionaryElementSyntax syntax, BindingDiagnosticBag diagnostics)
+            {
+                _ = BindValue(syntax.KeyExpression, diagnostics, BindValueKind.RValue);
+                _ = BindValue(syntax.ValueExpression, diagnostics, BindValueKind.RValue);
+                Error(diagnostics, ErrorCode.ERR_CollectionLiteralElementNotImplemented, syntax);
+                return BadExpression(syntax);
+            }
 
-                    case SpreadElementSyntax spreadElementSyntax:
-                        _ = BindValue(spreadElementSyntax.Expression, diagnostics, BindValueKind.RValue);
-                        Error(diagnostics, ErrorCode.ERR_CollectionLiteralElementNotImplemented, syntax);
-                        return BadExpression(syntax);
+            BoundExpression bindSpreadElement(SpreadElementSyntax syntax, BindingDiagnosticBag diagnostics)
+            {
+                var expression = BindRValueWithoutTargetType(syntax.Expression, diagnostics);
```

Comment:
Can you leave a note that this current doens't have a target type, but that it's an open spec question.  note that we likely want target typing to flow in for cases like:

```c#
ImmutableArray<string> x = [a, b, .. z ? [null] : []];
```

Without some sort of target typing flowing in, both those internal literals will be errors.
```

Spec mentions: > Do we need to target-type spread_element? Consider, for example:

Reply:
Added a PROTOTYPE comment to `BindCollectionInitializerCollectionLiteral()` and added `CollectionLiteralTests.SpreadElement_10`.